plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ptf/Query.java

```diff
@@ -0,0 +1,131 @@
+import io.trino.spi.ptf.AbstractConnectorTableFunction;
+import io.trino.spi.ptf.Argument;
+import io.trino.spi.ptf.ConnectorTableFunction;
+import io.trino.spi.ptf.ConnectorTableFunctionHandle;
+import io.trino.spi.ptf.Descriptor;
+import io.trino.spi.ptf.Descriptor.Field;
+import io.trino.spi.ptf.ScalarArgument;
+import io.trino.spi.ptf.ScalarArgumentSpecification;
+import io.trino.spi.ptf.TableFunctionAnalysis;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.spi.ptf.ReturnTypeSpecification.GenericTable.GENERIC_TABLE;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class Query
+        implements Provider<ConnectorTableFunction>
+{
+    public static final String NAME = "query";
+
+    private final JdbcTransactionManager transactionManager;
+
+    @Inject
+    public Query(JdbcTransactionManager transactionManager)
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+    }
+
+    @Override
+    public ConnectorTableFunction get()
+    {
+        return new ClassLoaderSafeConnectorTableFunction(new QueryFunction(transactionManager), getClass().getClassLoader());
+    }
+
+    public static class QueryFunction
+            extends AbstractConnectorTableFunction
+    {
+        public static final String SCHEMA_NAME = "system";
+
+        private final JdbcTransactionManager transactionManager;
+
+        public QueryFunction(JdbcTransactionManager transactionManager)
+        {
+            super(
+                    SCHEMA_NAME,
+                    NAME,
+                    List.of(ScalarArgumentSpecification.builder()
+                            .name("QUERY")
+                            .type(VARCHAR)
+                            .build()),
+                    GENERIC_TABLE);
+            this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        }
+
+        @Override
+        public TableFunctionAnalysis analyze(ConnectorSession session, ConnectorTransactionHandle transaction, Map<String, Argument> arguments)
+        {
+            ScalarArgument argument = (ScalarArgument) getOnlyElement(arguments.values());
+            String query = ((Slice) argument.getValue()).toStringUtf8();
+            PreparedQuery preparedQuery = new PreparedQuery(query, ImmutableList.of());
+
+            JdbcMetadata metadata = transactionManager.getMetadata(transaction);
+            JdbcTableHandle tableHandle = metadata.getTableHandle(session, preparedQuery);
+            List<JdbcColumnHandle> columns = tableHandle.getColumns().orElseThrow(() -> new IllegalStateException("Query result has no columns"));
```

Comment:
- no columns may mean "lack of column info" or "zero number of columns", let's be explicit
- Query result -> handle

"Handle doesn't have columns info"